system/mp/mpbase_mpi.cpp

```diff
@@ -0,0 +1,266 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#define mp_decl DECL_EXPORT
+#include "platform.h"
+#include "jlib.hpp"
+#include "jlog.hpp"
+
+#include "mpbase.hpp"
+#include "mpi_wrapper.hpp"
+
+static INode *MyNode=NULL;
+static INode *NullNode=NULL;
+
+
+IGroup *createIGroup(rank_t num,INode **nodes){
+    return new NodeGroup();
+}
+
+IGroup *createIGroup(rank_t num,const SocketEndpoint *ep){
+    UNIMPLEMENTED; 
+}
+
+IGroup *createIGroup(SocketEndpointArray &epa){
+    UNIMPLEMENTED; 
+}
+
+IGroup *createIGroup(const char *endpointlist,unsigned short defport){
+    UNIMPLEMENTED; 
+}
+
+class MPNode: implements INode, public CInterface
+{
+protected: friend class MPNodeCache;
+    SocketEndpoint ep;
+public:
+    IMPLEMENT_IINTERFACE;
+    MPNode(const SocketEndpoint &_ep)
+        : ep(_ep)
+    {
+#ifdef _DEBUG
+//      assertex(!_ep.LoopBack()); // localhost not supported for nodes
+#endif
+    }
+    bool equals(const INode *n) const { return endpoint().equals(n->endpoint()); }
+    void serialize(MemoryBuffer &tgt)
+    {
+        ep.serialize(tgt);
+    }
+    static MPNode *deserialize(MemoryBuffer &src);
+    unsigned getHash() const { return ep.hash(0); }
+    virtual bool isHost() const
+    {
+        return ep.isHost();
+    }
+    virtual bool isLocalTo(INode *to) const
+    {
+        return ep.ipequals(to->endpoint());
+    }
+    const SocketEndpoint &endpoint() const { return ep; }
+
+};
+
+class MPNodeCache: public SuperHashTableOf<MPNode,SocketEndpoint>
+{
+    CriticalSection sect;
+public:
+    ~MPNodeCache()
+    {
+        _releaseAll();
+    }
+
+    void onAdd(void *)
+    {
+        // not used
+    }
+
+    void onRemove(void *e)
+    {
+        MPNode &elem=*(MPNode *)e;      
+        elem.Release();
+    }
+
+    unsigned getHashFromElement(const void *e) const
+    {
+        const MPNode &elem=*(const MPNode *)e;      
+        return elem.ep.hash(0);
+    }
+
+    unsigned getHashFromFindParam(const void *fp) const
+    {
+        return ((const SocketEndpoint *)fp)->hash(0);
+    }
+
+    const void * getFindParam(const void *p) const
+    {
+        const MPNode &elem=*(const MPNode *)p;      
+        return &elem.ep;
+    }
+
+    bool matchesFindParam(const void * et, const void *fp, unsigned) const
+    {
+        return ((MPNode *)et)->ep.equals(*(SocketEndpoint *)fp);
+    }
+
+    IMPLEMENT_SUPERHASHTABLEOF_REF_FIND(MPNode,SocketEndpoint);
+
+    MPNode *lookup(const SocketEndpoint &ep)
+    {
+        CriticalBlock block(sect);
+        MPNode *item=SuperHashTableOf<MPNode,SocketEndpoint>::find(&ep);
+        if (!item) {
+            item = new MPNode(ep);
+            add(*item);
+        }
+        return LINK(item);
+    }
+
+
+} *NodeCache = NULL;
+
+MPNode *MPNode::deserialize(MemoryBuffer &src)
```

Comment:
Everything in mpbase_mpi.cpp is from the original mpbase.cpp. While implementing MPI stuff I did not come across working with an INode since we always dealt with rank or socketendpoint for direct communication. But however I think the functions I've ignored implementing for now dealt with JSocket/MPI communication independent executions. So I don't think they would differ at all.

Reply:
ok, they are prob. not effecting you as much as the reviewers ;) - but if there is code you're not using it would be easier for us at least if it was not copied. 
And where you need a slightly different version of something and it isn't too tricky, better to turn the existing implementation into base classes in hpp's and derive them, so we can really see what's new/different/required.